torch/__init__.py

```diff
@@ -778,3 +786,154 @@ def _register_device_module(device_type, module):
+            # torch.mm again runs in float16 and produces float16 output, regardless of input types.
+            g_float16 = torch.mm(d_float32, f_float32)
+    The autocast state is thread-local.  If you want it enabled in a new thread, the context manager or decorator
+    must be invoked in that thread.  This affects :class:`torch.nn.DataParallel` and
+    :class:`torch.nn.parallel.DistributedDataParallel` when used with more than one GPU per process
+    (see :ref:`Working with Multiple GPUs<amp-multigpu>`).
+    Args:
+        enabled(bool, optional, default=True):  Whether autocasting should be enabled in the region.
+    """
+    def __init__(self, enabled=True, fast_dtype=torch.float16, device='cuda'):
+        self.device = device
+        if device == 'cpu':
+            supported_dtype = [torch.bfloat16]
+            if fast_dtype not in supported_dtype :
+                warnings.warn("In CPU autocast, but the target dtype is not supported. Disable the autocast.")
+                warnings.warn("CPU Autocast only support dtype of torch.bfloat16 currently.")
+                enabled = False
+                fast_dtype = torch.bfloat16
+        self._enabled = enabled
+        self.fast_dtype = fast_dtype
+
+    def __enter__(self):
+        if self.device == 'cpu':
+            self.prev = torch.is_autocast_cpu_enabled()
+            self.prev_fastdtype = torch.get_autocast_cpu_dtype()
+            torch.set_autocast_cpu_enabled(self._enabled)
+            torch.set_autocast_cpu_dtype(self.fast_dtype)
+            torch.autocast_increment_nesting()
+        else:
+            self.prev = torch.is_autocast_enabled()
+            self.prev_fastdtype = torch.get_autocast_gpu_dtype()
+            torch.set_autocast_gpu_dtype(self.fast_dtype)
+            torch.set_autocast_enabled(self._enabled)
+            torch.autocast_increment_nesting()
+
+    def __exit__(self, *args):
+        # Drop the cache when we exit to a nesting level that's outside any instance of autocast.
+        if self.device == 'cpu':
+            if torch.autocast_decrement_nesting() == 0:
+                torch.clear_autocast_cache()
+            torch.set_autocast_cpu_enabled(self.prev)
+            torch.set_autocast_cpu_dtype(self.prev_fastdtype)
+        else:
+            if torch.autocast_decrement_nesting() == 0:
+                torch.clear_autocast_cache()
+            torch.set_autocast_enabled(self.prev)
+            torch.set_autocast_gpu_dtype(self.prev_fastdtype)
+        return False
+
+    def __call__(self, func):
+        @functools.wraps(func)
+        def decorate_autocast(*args, **kwargs):
+            with self:
+                return func(*args, **kwargs)
+        return decorate_autocast
+
+
+# Casts Tensors and containers of Tensors.  Special-cases passthroughs for strings and np.ndarrays, which
+# may be falsely detected as "Iterables."
+def _cast(value, dtype):
```

Comment:
this `_cast` function is not used anywhere in this PR. In torch/cuda/amp/autocast_mode.py `_cast` is used in `custom_fwd`, but you don't have `custom_fwd` here. Do you need `custom_fwd` here? If not, you don't need `_cast`